lib/semantics/check-do-stmt.cc

```diff
@@ -344,41 +372,101 @@ static CS GatherReferencesFromExpression(const parser::Expr &expression) {
   }
 }
 
-// Find a canonical DO CONCURRENT and enforce semantics checks on its body
-class DoConcurrentContext {
+// Find a DO statement and enforce semantics checks on its body
+class DoStmtContext {
 public:
-  DoConcurrentContext(SemanticsContext &context)
-    : messages_{context.messages()} {}
-
-  bool operator==(const DoConcurrentContext &x) const { return this == &x; }
+  DoStmtContext(SemanticsContext &context) : context_{context} {}
 
   void Check(const parser::DoConstruct &doConstruct) {
+    if (doConstruct.IsDoConcurrent()) {
+      CheckDoConcurrent(doConstruct);
+      return;
+    }
+    if (doConstruct.IsDoNormal()) {
+      CheckDoNormal(doConstruct);
+      return;
+    }
+    // TODO: handle the other cases
+  }
+
+private:
+  using Bounds = parser::LoopControl::Bounds;
+
+  const Bounds &GetBounds(const parser::DoConstruct &doConstruct) {
+    auto &loopControl{doConstruct.GetLoopControl().value()};
+    return std::get<Bounds>(loopControl.u);
+  }
+
+  void CheckDoControl(parser::CharBlock sourceLocation, bool isReal) {
+    bool warn = context_.warnOnNonstandardUsage() ||
+        context_.ShouldWarn(parser::LanguageFeature::RealDoControls);
+    bool enabled = context_.IsEnabled(parser::LanguageFeature::RealDoControls);
+    if (isReal && enabled && !(warn)) {
+      // No messages for the default case
+    } else if (isReal && enabled && warn && !context_.warningsAreErrors()) {
+      context_.Say(sourceLocation, "DO controls should be INTEGER"_en_US);
+    } else {
+      context_.Say(sourceLocation, "DO controls should be INTEGER"_err_en_US);
+    }
+  }
+
+  void CheckDoVariable(const parser::ScalarName &scalarName) {
+    const DeclTypeSpec *symType{scalarName.thing.symbol->GetType()};
```

Comment:
You should check that you actually got a type (`symType !=nullptr`). I can segfault this PR with the following program:


```
subroutine foo()
end subroutine

program test
  ! segfaults semantic checks
  do foo = 1, 10, 3
    print *, "segfault semantic checks"
  end do
end program
```
This is because some symbols, such as subroutine, have no type.

Also, you should probably check that `scalarName.thing.symbol` is indeed a varaiable (and also not a procedure pointer). The following program goes through semantic when it probably should not:
```
function ibar()
end function

program test
  procedure(ibar), pointer :: ibar2 => NULL()

  ! Is accepted when it should not
  do ibar = 1, 10, 3
    print *, "something"
  end do
  ! Same
  Do ibar2 = 1, 10, 3
    print *, "something"
  end do
end program
```

Reply:
Wow!  Thanks for catching this.  I'll add a fix and a test.